src/libraries/System.Private.CoreLib/src/System/Buffers/Binary/ReaderLittleEndian.cs

```diff
@@ -141,10 +180,26 @@ public static bool TryReadInt64LittleEndian(ReadOnlySpan<byte> source, out long
             return success;
         }
 
+        /// <summary>
+        /// Reads a Single out of a read-only span of bytes as little endian.
+        /// </summary>
+        /// <returns>If the span is too small to contain a Single, return false.</returns>
+        public static bool TryReadSingleLittleEndian(ReadOnlySpan<byte> source, out float value)
+        {
+            if (!BitConverter.IsLittleEndian)
+            {
+                bool success = MemoryMarshal.TryRead(source, out int temp);
```

Comment:
The existing TryRead... methods fetch this into `out value` that does not seem like a good idea - it just makes it harder for the JIT to optimize. I agree that fetching this to temp is better.

It would be nice to fix the existing TryRead... methods to follow the same style (fetch into `temp`).

Reply:
I did a quick benchmark for `short` and `long`, and the existing code seems to perform better on my machine:

```C#
    public class Benchmarks
    {
        private byte[] _buffer = new byte[] { 0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77, 0x88 };

        [Benchmark]
        public bool Int32Orig()
        {
            int value;
            return TryReadInt32BigEndianOrig(_buffer, out value);
        }

        [Benchmark]
        public bool Int32New()
        {
            int value;
            return TryReadInt32BigEndianNew(_buffer, out value);
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static bool TryReadInt32BigEndianOrig(ReadOnlySpan<byte> source, out int value)
        {
            bool success = MemoryMarshal.TryRead(source, out value);
            if (BitConverter.IsLittleEndian)
            {
                value = BinaryPrimitives.ReverseEndianness(value);
            }
            return success;
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static bool TryReadInt32BigEndianNew(ReadOnlySpan<byte> source, out int value)
        {
            if (BitConverter.IsLittleEndian)
            {
                bool success = MemoryMarshal.TryRead(source, out int temp);
                value = BinaryPrimitives.ReverseEndianness(temp);
                return success;
            }

            return MemoryMarshal.TryRead(source, out value);
        }

        [Benchmark]
        public bool Int16Orig()
        {
            short value;
            return TryReadInt16BigEndianOrig(_buffer, out value);
        }

        [Benchmark]
        public bool Int16New()
        {
            short value;
            return TryReadInt16BigEndianNew(_buffer, out value);
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static bool TryReadInt16BigEndianOrig(ReadOnlySpan<byte> source, out short value)
        {
            bool success = MemoryMarshal.TryRead(source, out value);
            if (BitConverter.IsLittleEndian)
            {
                value = BinaryPrimitives.ReverseEndianness(value);
            }
            return success;
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static bool TryReadInt16BigEndianNew(ReadOnlySpan<byte> source, out short value)
        {
            if (BitConverter.IsLittleEndian)
            {
                bool success = MemoryMarshal.TryRead(source, out short temp);
                value = BinaryPrimitives.ReverseEndianness(temp);
                return success;
            }

            return MemoryMarshal.TryRead(source, out value);
        }
    }
```
BenchmarkDotNet=v0.12.0, OS=Windows 10.0.18363
Intel Core i7-6700 CPU 3.40GHz (Skylake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.1.101
  [Host]     : .NET Core 3.1.1 (CoreCLR 4.700.19.60701, CoreFX 4.700.19.60801), X64 RyuJIT
  DefaultJob : .NET Core 3.1.1 (CoreCLR 4.700.19.60701, CoreFX 4.700.19.60801), X64 RyuJIT


|    Method |      Mean |     Error |    StdDev |    Median |
|---------- |----------:|----------:|----------:|----------:|
| Int32Orig | 0.6358 ns | 0.0576 ns | 0.1568 ns | 0.5749 ns |
|  Int32New | 0.9351 ns | 0.0484 ns | 0.0497 ns | 0.9484 ns |
| Int16Orig | 0.6282 ns | 0.0432 ns | 0.0383 ns | 0.6247 ns |
|  Int16New | 0.8781 ns | 0.0456 ns | 0.0381 ns | 0.8779 ns |

Reply:
```
  int value;
  return TryReadInt32BigEndianOrig(_buffer, out value);
```
You are not using the value for anything that is not realistic and likely skewing your results.

E.g. try to change the value to `static` so that the JIT cannot optimize it out - you should get pretty different results:
```
static int value;
...
  return TryReadInt32BigEndianOrig(_buffer, out value);
```

Reply:
@stephentoub Made the change for the existing methods already in https://github.com/dotnet/runtime/pull/20985

Reply:
> try to change the value to static so that the JIT cannot optimize it out

That's the test I'd done earlier today.  Recreating it now with the code you shared by with statics for the out arguments:

|    Method |      Mean |     Error |    StdDev |
|---------- |----------:|----------:|----------:|
| Int32Orig | 1.0450 ns | 0.0053 ns | 0.0045 ns |
|  Int32New | 0.5885 ns | 0.0026 ns | 0.0023 ns |
| Int16Orig | 1.0178 ns | 0.0017 ns | 0.0015 ns |
|  Int16New | 0.8376 ns | 0.0081 ns | 0.0072 ns |

Looking at the asm, the new versions are also identical when using a local and better when the indirection can't be eliminated.